docs: add intro to MUD (#296)
latticexyz/mud: guides/introduction.md

> # Introduction

Reviewer: it could be cool to include a gif of your increment system with hot reloading here (im a fan of gifs as the first thing people see)

  function remove(uint256 entity) external;

  function has(uint256 entity) external view returns (bool);

  function getRawValue(uint256 entity) external view returns (bytes memory);

  function getEntities() external view returns (uint256[] memory);

  function getEntitiesWithValue(bytes memory value) external view returns (uint256[] memory);

  function authorizeWriter(address writer) external;

  function unauthorizeWriter(address writer) external;

  function world() external view returns (address);
}

```

The base component implementation is extended by the typed component contracts, adding typed functions to set and get component values, for a more streamlined development process.

```solidity
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.0;

import { IComponent } from "./IComponent.sol";

interface IUint256Component is IComponent {
  function set(uint256 entity, uint256 value) external;

  function getValue(uint256 entity) external view returns (uint256);

  function getEntitiesWithValue(uint256 value) external view returns (uint256[] memory);
}

```

Components don't contain any logic besides the logic required to store and load component values.

> All application logic lives in systems, which don't contain any persistent state but work with state stored in components.

Reviewer 2: ```suggestion
All application logic lives in systems with no persistent state, and they work with state stored in components.
```

All systems can read from all components, but need write access to write to components.
(Write access is managed by a component's owner via the component's `authorizeWriter` and `unauthorizeWriter` functions.)

Note how multiple systems can share the same state and how there is no requirement for the systems and components to be deployed by the same developer.

At the center of every MUD application is the World contract.
The World is a global namespace, in which components and systems are registered with a fixed id.
The World contract is permissionless - anyone can register additional components and systems.
This is the most notable difference between this architecture and existing ones like the [Diamond pattern](https://eips.ethereum.org/EIPS/eip-2535).
It is possible because state is encapsulated in separate components contracts with individual write access.

> When a component value is changed, the update is registered on the World contract (via `registerComponentValueSet`).

Reviewer 2: ```suggestion
When a component value changes, the update is registered on the World contract (via `registerComponentValueSet`).
```

> This causes the World contract to emit an event (`ComponentValueSet`), which can be used by clients and indexers to mirror the entire World state.

Reviewer 2: ```suggestion
This causes the World contract to emit an event (`ComponentValueSet`), which clients and indexers can use to mirror the entire World state.
```

With `setupMUDNetwork` from the `@latticexyz/std-client` package, local RECS components can be linked to on-chain SOLECS components to synchronize their state.

```typescript
// From https://github.com/latticexyz/mud-template-minimal/blob/main/packages/client/src/index.ts
import { setupMUDNetwork, defineNumberComponent } from "@latticexyz/std-client";
import { createWorld } from "@latticexyz/recs";
import { SystemTypes } from "contracts/types/SystemTypes";
import { SystemAbis } from "contracts/types/SystemAbis.mjs";
import { config } from "./config";

const world = createWorld();

const components = {
  Counter: defineNumberComponent(world, {
    metadata: {
      // MUD will synchronize the local component with the
      // on chain component with id `keccak256("component.Counter")`.
      contractId: "component.Counter",
    },
  }),
};

const network = await setupMUDNetwork<typeof components, SystemTypes>(config, world, components, SystemAbis);
```

Note: the on-chain component is always considered the ground truth.
Setting a value on-chain will also update the local component, but setting a local value won't automatically update the on-chain value.

> Instead, the client should call an on-chain system, which modifies the on-chain component and thereby triggers an update of the local component.

Reviewer 2: ```suggestion
Instead, the client should call an on-chain system, which modifies the on-chain component and triggers an update of the local component.
```

```typescript
// Query for all entities with Counter component
const query = defineQuery([Has(Counter)]);
console.log("Entities with Counter:", query.matching);

// The query result includes a stream emitting new results
query.update$.subscribe((update) => console.log("Counter update:", update));

// Systems can react to query updates too (and much more, see RECS docs)
defineSystem(world, [Has(Counter)], (update) => {
  console.log("Counter update:", update);
});
```

## Further reading

MUD contains a few other packages besides the ones mentioned in this overview (SOLECS, Network, Services, RECS):

- `cli`: The MUD command line interface contains scripts to make the life of developers easier when working with MUD - scaffolding new projects, managing the development environment, contract hot reloading, contract deployment, type generation, etc.
- `ecs-browser`: A data explorer to visualize and edit on-chain and local component values - made possible by storing state in a standardized and introspectable way.
- `std-contracts`/`std-client`: MUD standard libraries for contract and client development.

> Now that you have a basic understanding of the core of MUD and the principles behind it, let's dive into a MUD project.

Reviewer 2: ```suggestion
Now that you have a basic understanding of the core of MUD and its principles, let's dive into a MUD project.
```
